File: push-openapi-python-demo/Custom_Signature/sign.py
import hashlib
import hmac
import base64
from popEncode import *

def sign(method,params,secret):
    query_string = ''
    for k in sorted(params.keys()):
        query_string += '&' + pop_encode(k) + "=" + pop_encode(params[k])
    sign_to_string = method + "&%2F&" + pop_encode(query_string[1:])
    h = hmac.new(secret, sign_to_string , hashlib.sha1)
    signature = base64.encodestring(h.digest())
    return signature
        

File: OpenApi1.0/push-openapi-nodejs-demo/Custom_Signature/sign.js
var cryptoLib =  require('crypto');
var popEscape = require('./popEscape.js');


module.exports = { 
sign : function(method, params, accessKeySecret) {
  var QueryString = '';
  var keys = Object.keys(params).sort();
  for (key in keys) {
    QueryString = QueryString + keys[key] + "=" + popEscape.popEscape(params[keys[key]]) + "&";
  }
  QueryString = QueryString.substring(0,QueryString.length-1)
  const stringToSign = method + "&%2F&" + popEscape.popEscape(QueryString);
  return hmac(accessKeySecret + '&', stringToSign, 'base64', 'sha1');
}
var hmac = function(key, string, digest, fn) {
  if (!digest) {
    digest = 'binary';
  }
  if (digest === 'buffer') {
    digest = undefined;
  }
  if (!fn) {
    fn = 'sha256';
  }
  if (typeof string === 'string') {
    string = new Buffer(string);
  }
  return cryptoLib.createHmac(fn, key).update(string).digest(digest);
}
};
